crates/sequencer/src/sequencer.rs

```diff
@@ -564,26 +429,123 @@ where
-                        error!("Sequencer error: {}", e);
-                    }
+        // Initialize our knowledge of the state of the DA-layer
+        let (mut prev_l1_height, mut last_finalized_block, mut l1_fee_rate) =
+            match get_da_block_data(self.da_service.clone(), self.ledger_db.clone()).await {
+                Ok(l1_data) => l1_data,
+                Err(e) => {
+                    error!("{}", e);
+                    return Err(e);
                 }
-            }
+            };
+
+        // Based on the above initialization, decide whether we should submit a commitment to DA.
+        if let Err(e) = self
+            .submit_commitment(last_finalized_block.header().height(), prev_l1_height)
+            .await
+        {
+            error!("Sequencer error: {}", e);
         }
-        // If sequencer is in production mode, it will build a block every 2 seconds
-        else {
-            loop {
-                sleep(Duration::from_secs(2)).await;
-                if let Err(e) = self.build_block().await {
-                    error!("Sequencer error: {}", e);
+
+        // Setup required workers to update our knowledge of the DA layer every 2 seconds.
+        let (da_tx, mut da_rx) = mpsc::channel(1);
+        let da_monitor = da_block_monitor(self.da_service.clone(), self.ledger_db.clone(), da_tx);
+        tokio::pin!(da_monitor);
+
+        let target_block_time = Duration::from_secs(2);
+        let mut parent_block_exec_time = Duration::from_secs(0);
+
+        // In case the sequencer falls behind on DA blocks, we need to produce at least 1
+        // empty block per DA block. Which means that we have to keep count of missed blocks
+        // and only resume normal operations once the sequencer has caught up.
+        let mut missed_da_blocks_count = 0;
+
+        loop {
+            let mut interval = tokio::time::interval(target_block_time - parent_block_exec_time);
+            // The first ticket completes immediately.
+            // See: https://docs.rs/tokio/latest/tokio/time/struct.Interval.html#method.tick
+            interval.tick().await;
+
+            let last_finalized_height = last_finalized_block.header().height();
+            tokio::select! {
+                // Run the DA monitor worker
+                _ = &mut da_monitor => {},
+                // Receive updates from DA layer worker.
+                l1_data = da_rx.recv() => {
+                    // Stop receiving updates from DA layer until we have caught up.
+                    if missed_da_blocks_count > 0 {
+                        continue;
+                    }
+                    if let Some(l1_data) = l1_data {
+                        (prev_l1_height, last_finalized_block, l1_fee_rate) = l1_data;
+
+                        if last_finalized_height > prev_l1_height {
+                            let skipped_blocks = last_finalized_height - prev_l1_height;
```

Comment:
let's declare skipped_blocks as last_finalized - prev_l1 - 1 

so that skipped blocks is word by word, the number of skipped blocks

and then we can set missed_da_blocks_count to skipped blocks directly.

e.g.

last_finalized = 5
prev_l1 = 2

skipped = 3, 4

skipped_blocka = 5-2-1 = 2